Missing event subject

ncounter opened the pull request:
Every subscribable `Event` has its own `subject` (for RSS and Email) implementation.

eduardoj commented on src/api/app/models/event/delete_package.rb:
```diff
@@ -4,6 +4,10 @@ class DeletePackage < Base
     payload_keys :project, :package, :sender, :comment, :requestid
 
+    def subject
+      "Package #{payload['package']} deleted"
```
```suggestion
    "Package #{payload['project']}/#{payload['package']} deleted"
```
I would add the project part too, like in other places (build_unchanged, build_success, ...).

hennevogel commented on src/api/app/models/event/build.rb:
```diff
@@ -5,6 +5,10 @@ class Build < Base
                  :rev, :reason, :bcnt, :verifymd5, :hostarch, :starttime, :endtime, :workerid, :versrel, :previouslyfailed, :successive_failcount, :buildtype
 
+    def subject
+      "Build finished for #{payload['project']}/#{payload['package']}"
```
We don't send emails or RSS entries for this.

eduardoj commented on src/api/app/models/event/repo_publish_state.rb:
```diff
@@ -3,6 +3,10 @@ class RepoPublishState < Base
     payload_keys :project, :repo, :state
+
+    def subject
+      "Repository #{payload[:repo]} state of #{payload[:project]} changed to #{payload[:project]}"
```
```suggestion
      "Repository #{payload[:repo]} state of #{payload[:project]} changed to #{payload[:state]}"
```
I guess this was a typo.

eduardoj commented on src/api/app/models/event/request_delete.rb:
```diff
@@ -2,6 +2,10 @@ module Event
     self.description = 'Request deleted (admin only)'
+
+    def subject
+      "Request #{payload['number']} deleted by #{payload['who']} (#{actions_summary})"
```
```suggestion
      "Request #{payload['number']} deleted by #{payload['who']}"
```
I'm not really sure if we should give many details about the actions of a request. I would prefer to show the details in the body of an email instead of in the subject.

hennevogel reviewed (changes requested):
You should check for which events we send emails/rss...

ncounter commented:
> You should check for which events we send emails/rss...

I thought about that but still, I wouldn't mind to have a fallback already implemented.
It's part of the interface they should implement by default to me.
Does that harm?

hennevogel commented:
> I thought about that but still, I wouldn't mind to have a fallback already implemented.
> It's part of the interface they should implement by default to me.

What interface? There is no interface to those that requires a subject. Or you mean the interface that might or might not exist in some future?

> Does that harm?

Sure. It's unused code without purpose. It will confuse people.

saraycp commented:
> I thought about that but still, I wouldn't mind to have a fallback already implemented.

In such case, I'd propose implementing the subject method in the parent class without a message but with: `raise AbstractMethodCalled`.

As we already do in other STI parent classes like [ScmPayload](https://github.com/openSUSE/open-build-service/blob/6c1c1684ff6e4a31e7abeebfae5a49972b0724cf/src/api/app/models/scm_payload.rb#L14).

ncounter commented:
> > I thought about that but still, I wouldn't mind to have a fallback already implemented.
> 
> In such case, I'd propose implementing the subject method in the parent class without a message but with: `raise AbstractMethodCalled`.
> 
> As we already do in other STI parent classes like [ScmPayload](https://github.com/openSUSE/open-build-service/blob/6c1c1684ff6e4a31e7abeebfae5a49972b0724cf/src/api/app/models/scm_payload.rb#L14).

That works for me :+1: 

@hennevogel , @saraycp 
After looking at how we send emails, in my understanding we do not filter events by any type in order to decide to send it or not. For instance, we do for web and RSS (see [here](https://github.com/openSUSE/open-build-service/blob/master/src/api/app/services/notification_service/notifier.rb#L54)), but there is nothing like that when sending email. So potentially all `Event`s can become an email, even if they are just meant to be STI parent classes of more specific event types.
I initially implemented a fallback subject text for STI parents, but raising the missing implementation sounds even better to me: it doesn't require maintenance and it would warn us in case we misuse `Event` classes.

I adapted the code accordingly. Feel free to take a look again. Thank you

hennevogel commented:
> So potentially all Events can become an email...

Nope, only the ones we allow to create subscriptions for...

hennevogel commented on src/api/app/models/event/repo_publish_state.rb:
```diff
@@ -3,6 +3,10 @@ class RepoPublishState < Base
     self.description = 'Publish state of repository changed'
     payload_keys :project, :repo, :state
+
+    def subject
```
We don't allow subscriptions for this event.

hennevogel commented on src/api/app/models/event/report_for_comment.rb:
```diff
@@ -3,6 +3,10 @@ class ReportForComment < Report
                  :project_name, :package_name, :commenter
+
+    def subject
+      'Comment reported'
```
```suggestion
      "Comment by #{payload[:commenter]} reported"
```

hennevogel commented on src/api/app/models/event/request_change.rb:
```diff
@@ -2,6 +2,10 @@ module Event
     self.message_bus_routing_key = 'request.change'
     self.description = 'Request XML updated (admin only)'
+
+    def subject
```
We don't allow subscriptions for this event.

hennevogel commented on src/api/app/models/event/request_delete.rb:
```diff
@@ -2,6 +2,10 @@ module Event
     self.message_bus_routing_key = 'request.delete'
     self.description = 'Request deleted (admin only)'
+
+    def subject
```
We don't allow subscriptions for this event.

hennevogel commented on src/api/app/models/event/status_check.rb:
```diff
@@ -2,5 +2,9 @@ module Event
     self.abstract_class = true
     payload_keys :who, :name, :short_description, :state, :url
+
+    def subject
```
We don't allow subscriptions for this event.

hennevogel commented on src/api/app/models/event/status_check_for_build.rb:
```diff
@@ -3,6 +3,10 @@ class StatusCheckForBuild < StatusCheck
     self.description = 'Status check for finished repository created'
     payload_keys :project, :repo, :arch, :buildid
+
+    def subject
```
We don't allow subscriptions for this event.

hennevogel commented on src/api/app/models/event/status_check_for_published.rb:
```diff
@@ -3,6 +3,10 @@ class StatusCheckForPublished < StatusCheck
     self.description = 'Status check for published repository created'
     payload_keys :project, :repo, :buildid
+
+    def subject
```
We don't allow subscriptions for this event.

hennevogel commented on src/api/app/models/event/status_check_for_request.rb:
```diff
@@ -2,6 +2,10 @@ module Event
     self.message_bus_routing_key = 'request.status_report'
     payload_keys :number
+
+    def subject
```
We don't allow subscriptions for this event.

hennevogel commented on src/api/app/models/event/undelete_package.rb:
```diff
@@ -4,6 +4,10 @@ class UndeletePackage < Base
     self.description = 'Package undeleted'
     payload_keys :project, :package, :sender, :comment
 
+    def subject
```
We don't allow subscriptions for this event.

hennevogel commented on src/api/app/models/event/undelete_project.rb:
```diff
@@ -3,6 +3,10 @@ class UndeleteProject < Base
     self.description = 'Project undeleted'
     payload_keys :project, :comment, :sender
+
+    def subject
```
We don't allow subscriptions for this event.

hennevogel commented on src/api/app/models/event/update_package.rb:
```diff
@@ -3,6 +3,10 @@ class UpdatePackage < Base
     self.description = 'Package meta data updated'
     payload_keys :project, :package, :sender
+
+    def subject
```
We don't allow subscriptions for this event.

hennevogel commented on src/api/app/models/event/update_project.rb:
```diff
@@ -3,6 +3,10 @@ class UpdateProject < Base
     self.description = 'Project meta updated'
     payload_keys :project, :sender
+
+    def subject
```
We don't allow subscriptions for this event.

hennevogel commented on src/api/app/models/event/update_project_config.rb:
```diff
@@ -3,6 +3,10 @@ class UpdateProjectConfig < Base
     self.description = 'Project _config updated'
     payload_keys :project, :sender, :files, :comment
+
+    def subject
```
We don't allow subscriptions for this event.

hennevogel commented on src/api/app/models/event/upload.rb:
```diff
@@ -3,6 +3,10 @@ class Upload < Base
     self.description = 'Package sources uploaded'
     payload_keys :project, :package, :sender, :comment, :filename, :requestid, :target, :user, :meta
+
+    def subject
```
We don't allow subscriptions for this event.

hennevogel commented on src/api/app/models/event/version_change.rb:
```diff
@@ -4,6 +4,10 @@ class VersionChange < Base
     self.description = 'Package changed its version'
     payload_keys :project, :package, :sender, :comment, :requestid, :files, :rev, :newversion, :user, :oldversion
 
+    def subject
```
We don't allow subscriptions for this event.

saraycp commented:
@hennevogel what do you mean by _"we don't allow subscriptions..."_. Is there any control/limitation in the code?
We don't create events for those classes but for their "children" classes, it seems it's a developer decision, I can't find the code that prevent us to do it.

ncounter commented:
@hennevogel I addressed all your comments by dropping the undesired `subject` implementation. (cc @saraycp )
Please take a final round of review, thank you

hennevogel commented on src/api/app/models/event/relationship_create.rb:
```diff
@@ -4,6 +4,10 @@ class RelationshipCreate < Relationship
     receiver_roles :any_role
+
+    def subject
+      "Relationship created with role #{payload[:any_role]}"
```
```suggestion
      object = payload[:project]
      object += "/#{payload[:package]} if payload[:package]
      "#{payload[:who]} added you as #{payload[role]} on #{object}"
```

hennevogel commented on src/api/app/models/event/relationship_delete.rb:
```diff
@@ -4,6 +4,10 @@ class RelationshipDelete < Relationship
     receiver_roles :any_role
+
+    def subject
+      "Relationship deleted with role #{payload[:any_role]}"
```
```suggestion
      object = payload[:project]
      object += "/#{payload[:package]} if payload[:package]
      "#{payload[:who]} added you as #{payload[role]} on #{object}"
```

hennevogel commented on src/api/app/models/event/report_for_package.rb:
```diff
@@ -3,6 +3,10 @@ class ReportForPackage < Report
     self.description = 'Report for a package created'
     payload_keys :package_name, :project_name
 
+    def subject
```
I'd love to see by **whom** here but that needs some refactoring first...
